baybe/utils/boolean.py

```diff
@@ -16,15 +16,17 @@ def is_abstract(cls: Any) -> bool:
     if a class has abstract methods. The latter can be problematic when the class has
     no abstract methods but is nevertheless not directly usable, for example, because it
     has uninitialized members, which are only covered in its non-"abstract" subclasses.
-    By contrast, this method simply checks if the class derives from ``abc.ABC``.
+
+    By contrast, this method simply checks if the class derives from ``abc.ABC`` or
+    is a protocol class.
 
     Args:
         cls: The class to be inspected.
 
     Returns:
         ``True`` if the class is "abstract" (see definition above), ``False`` else.
     """
-    return ABC in cls.__bases__
+    return (ABC in cls.__bases__) or (cls.__bases__ == (Protocol,))
```

Comment:
any reason for not writing `Protocol in cls.__bases__` ? Seems the logic is similar to `ABC` so why is the check different?

Reply:
I tested and think that your solution would work, too. I was just a bit overly careful here. But anyway, found an even better way. Just remembered that I saw something in the Python 3.13 release notes, an indeed there is now an `is_protocol` method that we can leverage via the `typing_extensions` backport. Nice side effect: this also fixed a "bug", namely we didn't have `typing_extensions` in our dependencies, even though we already use it for `TypeGuard`. So far, we were just lucky because we had it as an indirect dependency. Now resolved 😎 Please close if happy